Updated UI to fit new figma design. This also contains the new search…

sande11 opened the pull request:
# Pull Request: Enhanced Custom Action Bar with Integrated Search and Redesigned Bookmark Dialog

## **Overview**
This PR implements a comprehensive UI overhaul that matches Figma specifications. Link to design: https://www.figma.com/design/ZYZkodGKNJ3VOJMVlqdCMo/GA-TB-GUIDE-REDESIGN?node-id=2660-5103&m=dev

Before:
![before](https://github.com/user-attachments/assets/e72f8e97-822b-4b10-802e-d4ccfdb52295)

After:
<img width="1466" height="3101" alt="Screenshot_20250827_090157" src="https://github.com/user-attachments/assets/f59168f8-1c66-46d0-b6e7-b8bf8580c031" />


## **Key Features**

### 1. **Custom Action Bar System**
- **Dynamic Title Management**: Action bar titles dynamically update based on the current navigation destination
- **Smart Back Button**: Custom back arrow icon that appears/disappears contextually
- **Centered Title Layout**: All titles are properly centered with balanced spacing
- **Context-Aware Display**: Shows appropriate titles for chapters, sub-chapters, charts, and content pages

### 2. **Integrated Search Functionality**
- **Unified Search Design**: Search bar appears as part of the action bar for seamless integration
- **Fragment-Specific Search**: Search functionality only appears in content viewing screens (BodyFragment)

### 3. **Redesigned Bookmark Dialog**
- **Enhanced UX**: Improved button layouts, proper spacing, and intuitive clear functionality
- **Material Design Compliance**: Proper touch targets and accessibility features

## **Files Modified**

### **Core Architecture**
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/ActionBarController.kt` - New interface for action bar control
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/BaseFragment.kt` - Enhanced with action bar helper methods
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/MainActivity.kt` - Complete action bar management system

### **Layout Files**
- `app/src/main/res/layout/custom_action_bar_title.xml` - Redesigned custom action bar layout
- `app/src/main/res/layout/search_view.xml` - New integrated search view component
- `app/src/main/res/layout/fragment_body.xml` - Updated to include search integration
- `app/src/main/res/layout/dialog_bookmark.xml` - Complete Figma-based redesign

### **Drawable Resources**
- `app/src/main/res/drawable/search_view_background.xml` - Blue rectangular background for search container
- `app/src/main/res/drawable/frame_16.xml` - Gray background for search input field
- `app/src/main/res/drawable/rectangle_2.xml` - White dialog background with rounded corners
- `app/src/main/res/drawable/cancel_button_bg.xml` - Gray cancel button background
- `app/src/main/res/drawable/save_button_bg.xml` - Blue save button background
- `app/src/main/res/drawable/frame_46.xml` - Gray text field background for bookmark dialog

### **Fragment Updates**
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/fragments/BodyFragment.kt` - Search integration and title management
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/fragments/ChapterFragment.kt` - Removed search icon
- `app/src/main/java/org/apphatchery/gatbreferenceguide/ui/fragments/SubChapterFragment.kt` - Updated title handling and removed search icon

##  **Technical Improvements**

### **Action Bar Management**
```kotlin
interface ActionBarController {
    fun setActionBarTitle(title: String)
    fun setActionBarConfig(title: String, showBackButton: Boolean)
    fun setActionBarSearchVisible(visible: Boolean)
    fun setupActionBarSearch(onSearchAction: (String) -> Unit, onSearchIconClick: (String) -> Unit)
}
```

### **Dynamic Title System**
- **Chapter Navigation**: Shows full chapter names (e.g., "I. Epidemiology")
- **Content Pages**: Displays chart/table titles 
- **Bookmark Pages**: Shows "My Bookmarks" with custom back arrow
- **Navigation Context**: Proper title context for all navigation destinations

### **Search Integration**
- **Contextual Visibility**: Search only appears in content viewing screens
- **Performance Optimized**: Dynamic view finding prevents memory leaks

## **Design Specifications Met**

### **Bookmark Dialog Design**
- White background with 4dp rounded corners
- "Bookmark name" label positioned at 30dp from left and top
- Gray text field 
- Blue clear icon inside text field with proper tinting
- Full-width touching buttons 
- Cancel button: Gray background (`#ECECEC`)
- Save button: Blue background (`#5182FF`)

## **User Experience Improvements**

### **Navigation Flow**
1. **Consistent Title Display**: All screens show contextually appropriate titles
2. **Intuitive Back Navigation**: Smart back button appears only when needed
3. **Search Context**: Search functionality appears only in relevant content screens
4. **Visual Cohesion**: Action bar and search appear as unified design element

### **Bookmark Management**
1. **Improved Dialog UX**: Better visual hierarchy with proper spacing and colors
2. **Professional Appearance**: Matches design  with exact Figma specifications

## **Testing Considerations**
- **Navigation Testing**: Verify titles appear correctly on all navigation destinations
- **Bookmark Operations**: Ensure create, update, and delete operations work without crashes
- **Layout Testing**: Verify proper appearance across different screen sizes and orientations

## **Breaking Changes**
- **Custom Action Bar**: Replaces default Android action bar with custom implementation
- **Search Interface**: New search UI replaces previous implementation
- **Bookmark Dialog**: Complete redesign may affect existing bookmark workflows

sande11 committed:
Updated UI to fit new figma design. This also contains the new search view for body fragment

sande11 committed:
updated from API 34 to 35 and added edge-to-edge opt option

sande11 committed:
search functionality, bookmark page, search page

sande11 committed:
added feedback cards in body fragment, added bottom modal for notes with its floating button

sande11 committed:
implemented new design for the floating button

MaxwellKJr commented:
@sande11 great work! This looks solid. Just a few things that you need to look into:

### Align Chapter Cards text to the left
<img width="411" height="329" alt="Screenshot 2025-09-19 at 09 12 15" src="https://github.com/user-attachments/assets/29590c64-8ba6-462a-aa50-e98e11d42f6e" />

### Web View Search Options Not Fitting On Small Devices
- Try reducing the horizontal padding of the parent layout and also the individual elements
<img width="424" height="101" alt="Screenshot 2025-09-19 at 09 16 09" src="https://github.com/user-attachments/assets/92cd24f0-eba3-415b-aa8e-5ef0f6213aa5" />

### Clear Search Field When Last letter has been cleared with the keyboard delete button
<img width="430" height="99" alt="Screenshot 2025-09-19 at 09 27 05" src="https://github.com/user-attachments/assets/30f27f48-865c-408b-830c-b610cf166544" />
<img width="428" height="829" alt="Screenshot 2025-09-19 at 09 39 31" src="https://github.com/user-attachments/assets/c78ee26d-a95c-4c32-8637-3d63f298a3af" />

### Dismiss the Keyboard when popping out of the WebView page
<img width="430" height="867" alt="Screenshot 2025-09-19 at 09 38 48" src="https://github.com/user-attachments/assets/16989c5d-8e54-4d7f-b063-fbc5f5b1d7e4" />


### Dark Mode Not Working
[screen-20250919-090219.webm](https://github.com/user-attachments/assets/8ac9702f-903b-4ebc-b099-fb808602ea91)

MaxwellKJr reviewed (changes requested):
@sande11 great work! This looks solid. Just a few things that you need to look into:

### Align Chapter Cards text to the left
<img width="411" height="329" alt="Screenshot 2025-09-19 at 09 12 15" src="https://github.com/user-attachments/assets/29590c64-8ba6-462a-aa50-e98e11d42f6e" />

### Web View Search Options Not Fitting On Small Devices
- Try reducing the horizontal padding of the parent layout and also the individual elements
<img width="424" height="101" alt="Screenshot 2025-09-19 at 09 16 09" src="https://github.com/user-attachments/assets/92cd24f0-eba3-415b-aa8e-5ef0f6213aa5" />

### Clear Search Field When Last letter has been cleared with the keyboard delete button
<img width="430" height="99" alt="Screenshot 2025-09-19 at 09 27 05" src="https://github.com/user-attachments/assets/30f27f48-865c-408b-830c-b610cf166544" />
<img width="428" height="829" alt="Screenshot 2025-09-19 at 09 39 31" src="https://github.com/user-attachments/assets/c78ee26d-a95c-4c32-8637-3d63f298a3af" />

### Dismiss the Keyboard when popping out of the WebView page
<img width="430" height="867" alt="Screenshot 2025-09-19 at 09 38 48" src="https://github.com/user-attachments/assets/16989c5d-8e54-4d7f-b063-fbc5f5b1d7e4" />


### Dark Mode Not Working
[screen-20250919-090219.webm](https://github.com/user-attachments/assets/8ac9702f-903b-4ebc-b099-fb808602ea91)

sande11 committed:
fixed searchview responsiveness for small devices, fixed clear text logic in search view

sande11 committed:
fixed night mode crashes and implemented other night mode designs

sande11 committed:
fixed dismiss() keyboard and stacking card dialogs on top of each other bugs

sande11 committed:
Implemented Pendo tracking for note submission, added theme colors to Dialog_font text

sande11 committed:
bookmark card item margin start correction

sande11 committed:
Bug fixes: added new subchapters to fix missing content, fixed search results duplication bug, keyboard not being dismissed, search loading icon stuck on screen, bookmark card cut off, search result tabs being clipped

sande11 committed:
implemented: hide Note Floating button when note.size = 0

sande11 committed:
text accessibility

Changed the text in body fragment to expand when display and font size of the display changes

sande11 committed:
adjusted CSS to make tables 9,12,17,19 appear in the same font as the rest of the app and subchapter.json to remove a duplicate subchapter title

sande11 committed:
fixed highlighting in bookmark success pop up

sande11 commented:
@Wiza-Munthali

sande11 committed:
incremented BUILD_VERSION to 12 and Sept-24- to Oct-25-

private const val BUILD_VERSION = 12
private const val PENDO_RELEASE_VERSION = "Oct-25-"

sande11 committed:
Merge branch '131-update-tb-guide-android-app-ui-with-new-figma-design' into 133-update-android-app-to-target-latest-api-level-and-google-sdks

sande11 committed:
Merge pull request #136 from AppHatchery/133-update-android-app-to-target-latest-api-level-and-google-sdks

updated from API 34 to 35 (#133) in TB Guide android and then merged that branch into #131 that way we have both UI updates and API level updates in one branch

sande11 committed:
Revert "updated from API 34 to 35 in TB Guide android UI Updates"

sande11 committed:
Merge pull request #137 from AppHatchery/revert-136-133-update-android-app-to-target-latest-api-level-and-google-sdks

Revert "updated from API 34 to 35 in TB Guide android UI Updates"

sande11 committed:
improved bookmark and bookmark deletion card

sande11 committed:
Merge pull request #138 from AppHatchery/131-update-tb-guide-android-app-ui-with-new-figma-design

Merging #131 update-tb-guide-android-app-UI-with-new-figma-design into #133-update-android-app-to-target-latest-api-level-and-google-sdks

sande11 committed:
Update project configuration to API 36 and 16kb page compatibility

- Bump compileSdk and targetSdk to 36
- Update dependencies to latest versions
- Implement edge-to-edge layout in MainActivity
- Add status bar background view in activity_main.xml
- Add support for 16KB page size in gradle.properties

sande11 committed:
Refactor tab styles and structure

Updated tab container structure to improve appearance when font size is increased

sande11 committed:
Expand sections containing search results in BodyFragment

Added logic to automatically expand collapsed dropdowns, tabs, and details elements that contain search results before performing in-page search. Also added a method to collapse expanded sections when clearing the search. Updated custom_action_bar_title.xml to use wrap_content for height with a minimum height of 56dp for better layout flexibility.

sande11 committed:
Update floating notes button layout constraints moving it to the bottom

Changed the floating notes button margins and constraint to improve positioning. MarginEnd increased to 25dp, marginBottom added, and constraint now aligns to bottom_toolbar instead of search_view_include.

sande11 committed:
Merge pull request #146 from AppHatchery/131-update-tb-guide-android-app-ui-with-new-figma-design

Merge the latest UI updates to API level update branch

sande11 committed:
Refactor recent search removal logic

Replaces removeLast() with removeAt(recentSearches.lastIndex) to remove the last item from the recent searches list, ensuring compatibility

sande11 committed:
Refactor recent search removal logic

Replaces removeLast() with removeAt(recentSearches.lastIndex) to remove the last item from the recent searches list, ensuring compatibility

sande11 committed:
Merge pull request #147 from AppHatchery/133-update-android-app-to-target-latest-api-level-and-google-sdks

133 update android app to target latest api level and google sdks

sande11 committed:
Update app launcher icons and background assets

Replaced launcher icons and round icons for all densities, updated ic_launcher_background.xml with a new vector design, and added ic_launcher_foreground assets to mipmap folders. Updated adaptive icon XMLs to reference foreground images from mipmap instead of drawable for proper icon rendering.

sande11 committed:
Update external links and chapter references in HTML pages

Corrected URLs for CDC TB case definitions and QuantiFERON-TB Gold Plus package insert. Updated chapter references in HIV therapy tables to point to the correct section. These changes improve accuracy and navigation for users.

sande11 committed:
Update chapter links in HIV/TB therapy tables

Corrected the 'View in chapter' links in three HTML table pages to point to the appropriate chapter without fragment identifiers, ensuring consistency and proper navigation for users.

sande11 committed:
Merge pull request #148 from AppHatchery/144-chapter-link-error

144 Fix Broken internal and External Links

sande11 committed:
Improve bottom navigation behavior and layout padding

Added custom selection and reselection handlers for bottom navigation to ensure consistent navigation to root fragments. Updated fragment_saved.xml to include bottom padding for navigation bar, and defined bottom_nav_height in dimens.xml.

sande11 committed:
Update TB HTML pages for improved formatting and accuracy

Corrected image paths, improved table formatting, fixed HTML entities, and updated external links in multiple TB-related asset pages. These changes enhance clarity, consistency, and ensure accurate representation of diagnostic and treatment criteria.

sande11 committed:
Merge pull request #149 from AppHatchery/142-guide-nav-button-on-bookmark-screen-doesnt-work

 Fix bottom-nav interaction on Bookmarks screen

sande11 committed:
139 large font error:  WebView Font Persistence & Chapter 5C Layout Shift Fix + Dynamic Scaling for Chapter Icon and Paragraph Bar (#150)

* Comment out onZoomOut() call in WebView setup

The onZoomOut() method call in setupWebView was commented out, to prevent automatic zooming behavior which was causing conflicts in font size persistence and causing content to be pushed to the left of the screen

* Dynamically scale chapter icons with font size

Added logic to compute and apply chapter icon size in the WebView based on the current font size. This includes setting a CSS variable, tagging relevant images, injecting override styles, and applying inline sizing to ensure consistent scaling even if external styles are missing. The update is triggered after layout and on every page load.

* Scale decorative line in paragraphs with text zoom

Adds logic to dynamically scale the `.uk-paragraph::before` decorative line in response to text zoom, ensuring consistent appearance. The computed CSS values are injected via a style element in the WebView.

sande11 committed:
Fix font scaling and update bookmark dialog layout (#152)

Override attachBaseContext in App and MainActivity to enforce a fontScale of 1.0, preventing system font size changes from affecting the app. Simplify dialog_bookmark_success.xml by replacing the ScrollView with a single TextView and disabling auto-sizing for consistent appearance.

sande11 commented:
closed by #161
